### PR TITLE
[IMP] Log mismatch of --proxy-mode and X-Forwarded-Host header.

### PR DESCRIPTION
**Pull request by amh-mw**
This logs a warning to help administrators identify a misconfigured proxy upstream that is failing to provide X-Forwarded-Host when --proxy-mode is set.

This is a small follow-on to https://github.com/odoo/documentation/pull/6729 to grant a second chance to notice the problem in case the documentation wasn't read or understood.

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by amh-mw** [comment]
@Julien00859 viz

File: odoo/service/wsgi_server.py [hunk 1]
```diff
@@ -106,7 +106,9 @@ def application(environ, start_response):
...
     #        it
-    if config['proxy_mode'] and 'HTTP_X_FORWARDED_HOST' in environ:
-        return ProxyFix(application_unproxied)(environ, start_response)
-    else:
-        return application_unproxied(environ, start_response)
+    if config['proxy_mode']:
+        if 'HTTP_X_FORWARDED_HOST' in environ:
+            return ProxyFix(application_unproxied)(environ, start_response)
+        else:
+            _logger.warning("X-Forwarded-Host missing from header; --proxy-mode ignored")
```
Comment (Julien00859) [inline odoo/service/wsgi_server.py:113]:
```suggestion
            _logger.warning("X-Forwarded-Host missing from header; --proxy-mode ignored for this request")
```

**Review by Julien00859 — COMMENTED** [review]
> Not to shoot my own pull request in the foot, but it occurs to me that this could be unreasonably loud for the situation where not all requests come in through a proxy -- i.e. public traffic via proxy, private traffic via WAN / VPN.

I've a mixed feeling about adding a warning... On the one hand it is very important that people who deploy Odoo in production configure their proxies and use them to route all requests (so VPN -> nginx -> Odoo and not VPN -> Odoo). On the other hand in a dev environment this warning is going to be super annoying, e.g. on my own setup I always use `--proxy-mode` (it's in my config file) but when I run the test suite all the HttpCase connects directly to localhost:8069.

> I considered adding a global warning fuse like _flanker_lib_warning for proxy mode, but generally think that it's better to loudly complain if an upstream proxy isn't behaving as expected.

I too think that we should warn every single time, I don't think that logging the message only once applies here.

Maybe we could do instead:
```py
_logger_proxy = _logger.getChild("proxy")
_logger_proxy.warning(...)
```
this way we can add a configuration to just ignore that logger via `--log-handler odoo.service.wsgi_server.proxy=ERROR` (also possible via `.odoorc`)

If we go with this change, I would open the PR in 16.0 as ProxyFix as been moved to `odoo.http`. Or I would `_logger_proxy = logging.getLogger('odoo.http.proxy')` here in 15.0 for the sake of using the same logger name.

What do you think?

**Comment by amh-mw** [comment]
> What do you think?

I think that all sounds reasonable and have implemented it as described.

File: odoo/service/wsgi_server.py [hunk 2]
```diff
@@ -22,6 +22,7 @@
 from odoo.tools import config
 
 _logger = logging.getLogger(__name__)
+_logger_proxy = logging.getLogger('odoo.http.proxy')
```
Comment (Julien00859) [inline odoo/service/wsgi_server.py:25]:
```suggestion
_logger_proxy = logging.getLogger('odoo.http.proxy')  # forward compatibility with 16.0
```

**Review by Julien00859 — COMMENTED** [review]
Can you tell people about `--log-handler` and also reference the 16.0 PR inside the commit message?
@fw-bot up to saas-15.2

**Comment by Julien00859** [comment]
Going to `/` with the header set:
```
44:08.387 INFO werkzeug: 127.0.0.1 - - "GET /web HTTP/1.0" 303 - 2 0.016 0.120
44:09.404 INFO werkzeug: 127.0.0.1 - - "GET /web/service-worker.js HTTP/1.0" 404 - 2 0.001 0.005
44:10.382 INFO werkzeug: 127.0.0.1 - - "GET /web/login HTTP/1.0" 200 - 96 0.051 1.922
44:10.520 INFO werkzeug: 127.0.0.1 - - "GET /web/assets/188-a90a81e/web.assets_common.min.css HTTP/1.0" 200 - 5 0.002 0.017
44:10.528 INFO werkzeug: 127.0.0.1 - - "GET /web/assets/189-64fd00c/web.assets_frontend.min.css HTTP/1.0" 200 - 3 0.005 0.020
44:10.529 INFO werkzeug: 127.0.0.1 - - "GET /web/binary/company_logo HTTP/1.0" 200 - 2 0.003 0.018
44:10.530 INFO werkzeug: 127.0.0.1 - - "GET /web/assets/190-94175eb/web.assets_common_minimal.min.js HTTP/1.0" 200 - 3 0.005 0.021
44:10.532 INFO werkzeug: 127.0.0.1 - - "GET /web/assets/191-1bed1ae/web.assets_frontend_minimal.min.js HTTP/1.0" 200 - 3 0.003 0.021
44:10.699 INFO werkzeug: 127.0.0.1 - - "GET /web/assets/192-c9cbdc5/web.assets_common_lazy.min.js HTTP/1.0" 200 - 3 0.001 0.009
44:10.865 INFO werkzeug: 127.0.0.1 - - "GET /web/assets/193-fabd3ee/web.assets_frontend_lazy.min.js HTTP/1.0" 200 - 3 0.001 0.008
44:11.016 INFO werkzeug: 127.0.0.1 - - "GET / HTTP/1.0" 303 - 2 0.002 0.006
44:11.026 INFO werkzeug: 127.0.0.1 - - "GET /web/webclient/translations/1701708250930 HTTP/1.0" 200 - 7 0.009 0.008
44:11.033 INFO werkzeug: 127.0.0.1 - - "GET /web HTTP/1.0" 303 - 2 0.002 0.006
44:11.065 INFO werkzeug: 127.0.0.1 - - "GET /web/login HTTP/1.0" 200 - 6 0.005 0.018
44:11.094 INFO werkzeug: 127.0.0.1 - - "GET /web/webclient/qweb/1701708251032?bundle=web.assets_frontend HTTP/1.0" 200 - 10 0.006 0.043
44:11.115 INFO werkzeug: 127.0.0.1 - - "GET /web/webclient/translations/1701708251102 HTTP/1.0" 200 - 4 0.001 0.005
44:11.447 INFO werkzeug: 127.0.0.1 - - "GET /web/service-worker.js HTTP/1.0" 404 - 2 0.001 0.003
```
without:
```
44:15.348 WARNING o.h.proxy: X-Forwarded-Host missing from header; --proxy-mode ignored for this request 
44:15.357 INFO werkzeug: 127.0.0.1 - - "GET / HTTP/1.1" 303 - 2 0.002 0.006
44:15.370 WARNING o.h.proxy: X-Forwarded-Host missing from header; --proxy-mode ignored for this request 
44:15.373 INFO werkzeug: 127.0.0.1 - - "GET /web HTTP/1.1" 303 - 1 0.000 0.002
44:15.381 WARNING o.h.proxy: X-Forwarded-Host missing from header; --proxy-mode ignored for this request 
44:15.396 INFO werkzeug: 127.0.0.1 - - "GET /web/login HTTP/1.1" 200 - 5 0.004 0.011
44:15.470 WARNING o.h.proxy: X-Forwarded-Host missing from header; --proxy-mode ignored for this request 
44:15.471 WARNING o.h.proxy: X-Forwarded-Host missing from header; --proxy-mode ignored for this request 
44:15.474 WARNING o.h.proxy: X-Forwarded-Host missing from header; --proxy-mode ignored for this request 
44:15.475 INFO werkzeug: 127.0.0.1 - - "GET /web/static/lib/fontawesome/fonts/fontawesome-webfont.woff2?v=4.7.0 HTTP/1.1" 200 - - - -
44:15.476 WARNING o.h.proxy: X-Forwarded-Host missing from header; --proxy-mode ignored for this request 
44:15.480 WARNING o.h.proxy: X-Forwarded-Host missing from header; --proxy-mode ignored for this request 
44:15.492 INFO werkzeug: 127.0.0.1 - - "GET /web/assets/188-a90a81e/web.assets_common.min.css HTTP/1.1" 200 - 3 0.004 0.016
44:15.496 INFO werkzeug: 127.0.0.1 - - "GET /web/assets/189-64fd00c/web.assets_frontend.min.css HTTP/1.1" 200 - 3 0.005 0.017
44:15.497 INFO werkzeug: 127.0.0.1 - - "GET /web/assets/190-94175eb/web.assets_common_minimal.min.js HTTP/1.1" 200 - 3 0.002 0.018
44:15.498 INFO werkzeug: 127.0.0.1 - - "GET /web/assets/191-1bed1ae/web.assets_frontend_minimal.min.js HTTP/1.1" 200 - 3 0.005 0.012
44:15.499 WARNING o.h.proxy: X-Forwarded-Host missing from header; --proxy-mode ignored for this request 
44:15.503 INFO werkzeug: 127.0.0.1 - - "GET /web/binary/company_logo HTTP/1.1" 200 - 2 0.001 0.003
44:15.540 WARNING o.h.proxy: X-Forwarded-Host missing from header; --proxy-mode ignored for this request 
44:15.540 INFO werkzeug: 127.0.0.1 - - "GET /web_enterprise/static/img/home-menu-bg-overlay.svg HTTP/1.1" 200 - - - -
44:15.556 WARNING o.h.proxy: X-Forwarded-Host missing from header; --proxy-mode ignored for this request 
44:15.557 INFO werkzeug: 127.0.0.1 - - "GET /web_enterprise/static/fonts/Roboto/Roboto-Regular-webfont.woff HTTP/1.1" 200 - - - -
44:15.584 WARNING o.h.proxy: X-Forwarded-Host missing from header; --proxy-mode ignored for this request 
44:15.585 INFO werkzeug: 127.0.0.1 - - "GET /web/static/img/favicon.ico HTTP/1.1" 200 - - - -
44:15.633 WARNING o.h.proxy: X-Forwarded-Host missing from header; --proxy-mode ignored for this request 
44:15.643 INFO werkzeug: 127.0.0.1 - - "GET /web/assets/192-c9cbdc5/web.assets_common_lazy.min.js HTTP/1.1" 200 - 3 0.001 0.008
44:15.757 WARNING o.h.proxy: X-Forwarded-Host missing from header; --proxy-mode ignored for this request 
44:15.763 INFO werkzeug: 127.0.0.1 - - "GET /web/assets/193-fabd3ee/web.assets_frontend_lazy.min.js HTTP/1.1" 200 - 3 0.001 0.005
44:15.828 WARNING o.h.proxy: X-Forwarded-Host missing from header; --proxy-mode ignored for this request 
44:15.836 INFO werkzeug: 127.0.0.1 - - "GET /web/webclient/translations/1701708255824 HTTP/1.1" 200 - 4 0.002 0.005
44:15.843 WARNING o.h.proxy: X-Forwarded-Host missing from header; --proxy-mode ignored for this request 
44:15.880 INFO werkzeug: 127.0.0.1 - - "GET /web/webclient/qweb/1701708255840?bundle=web.assets_frontend HTTP/1.1" 200 - 10 0.002 0.033
44:15.890 WARNING o.h.proxy: X-Forwarded-Host missing from header; --proxy-mode ignored for this request 
44:15.898 INFO werkzeug: 127.0.0.1 - - "GET /web/webclient/translations/1701708255886 HTTP/1.1" 200 - 4 0.001 0.005
```
It is too spammy :( I know people are going to complain **a lot** (also at work) and I'm no more sure this is the right thing to do

**Comment by Julien00859** [comment]
> > It is too spammy :( I know people are going to complain **a lot** (also at work) even if there is a way to opt-out and I'm no more sure this is the right thing to do.
> 
> I could make the fuse time based? Something like one message every ten minutes? I will never see this myself, as I run an nginx container alongside my development odoo container and my test container doesn't set --proxy-mode.

This is all too complicated for a single log line, we have to keep it simple stupid inside the source code, using another dedicated logger is the most I'm willing to see implemented. Also imagine the situation where you (wrongly) changed smth inside your nginx configuration, a simple request wouldn't tell you if the new config is right, you would have to wait 10 minutes.

**Pull request closed without merging**